app/controllers/composer_controller.rb

```diff
@@ -0,0 +1,165 @@
+# frozen_string_literal: true
+
+class ComposerController < ApplicationController
+  requires_login
+
+  def mentions
+    names = params.require(:names)
+    raise Discourse::InvalidParameters.new(:names) if !names.kind_of?(Array) || names.size > 20
+
+    if params[:topic_id].present?
+      topic = Topic.find_by(id: params[:topic_id])
+      guardian.ensure_can_see!(topic)
+    end
+
+    # allowed_names is necessary just for new private messages.
+    allowed_names = if params[:allowed_names].present?
+      params[:allowed_names].split(',') << current_user.username
+    else
+      []
+    end
+
+    user_reasons = {}
+    group_reasons = {}
+    here_count = nil
+
+    users = User
+      .where(username_lower: names.map(&:downcase), staged: false)
+      .index_by(&:username_lower)
+
+    groups = Group
+      .visible_groups(current_user)
+      .where(name: names)
```

Comment:
Just curious, do we need to lowercase the name here?

Reply:
We do not have something like username_lower for users and cannot efficiently lookup groups, but I guess performance is not a concern here (we also do compare `lower(name)` in a few other places).

Reply:
I think we still need to search for groups by lower  casing the names.